src/advancedHMC_MCMC.jl

```diff
@@ -0,0 +1,317 @@
+    f.prob.u0 .+ (t .- f.prob.tspan[1]) .* y
+end
+
+# ODE DU/DX
+function NNodederi(phi::LogTargetDensity, t::AbstractVector, θ, autodiff::Bool)
+    if autodiff
+        # returns matrix [derivative returns vector(vector)]
+        hcat(ForwardDiff.derivative.(ti -> phi(ti, θ), t)...)
+    else
+        (phi(t .+ sqrt(eps(eltype(t))), θ) - phi(t, θ)) ./ sqrt(eps(eltype(t)))
+    end
+end
+
+# physloglike over problem timespan
+function physloglikelihood(Tar::LogTargetDensity, θ)
+    p = Tar.prob.p
+    f = Tar.prob.f
+
+    allparams = Tar.priors
+    invparams = allparams[2:length(allparams)]
+    meaninv = [invparam[1] for invparam in invparams]
+
+    autodiff = Tar.autodiff
+    dt = Tar.physdt
+    t = collect(Float64, Tar.prob.tspan[1]:dt:Tar.prob.tspan[2])
+
+    # # compare derivatives(matrix)
+    out = Tar(t, θ[1:(length(θ) - Tar.extraparams)])
+
+    # # this is a vector{vector{dx,dy}}(handle case single u(float passed))
+    if length(out[:, 1]) == 1
+        # shifted by prior mean
+        ode_params = exp.(θ[((length(θ) - Tar.extraparams) + 1):length(θ)] + log.(meaninv))
+        physsol = [f(out[:, i][1],
+                     ode_params,
+                     t[i])
+                   for i in 1:length(out[1, :])]
+    else
+        # shifted by prior mean
+        ode_params = exp.(θ[((length(θ) - Tar.extraparams) + 1):length(θ)] + log.(meaninv))
+        physsol = [f(out[:, i],
+                     ode_params,
+                     t[i])
+                   for i in 1:length(out[1, :])]
+    end
+    physsol = hcat(physsol...)
+
+    # # convert to matrix as nnsol
+    nnsol = NNodederi(Tar, t, θ[1:(length(θ) - Tar.extraparams)], autodiff)
+
+    physlogprob = 0
+    n = length(out[1, :])
+    for i in 1:length(Tar.prob.u0)
+        # can add phystd[i] for u[i]
+        physlogprob += logpdf(MvNormal(nnsol[i, :], Tar.phystd[i]), physsol[i, :])
+    end
+    return physlogprob
+end
+
+# Standard L2 losses training dataset
+function L2LossData(Tar::LogTargetDensity, θ)
+    # matrix(each row corresponds to vector u's rows)
+    nn = Tar(Tar.dataset[length(Tar.dataset)], θ[1:(length(θ) - Tar.extraparams)])
+
+    L2logprob = 0
+    n = length(nn[1, :])
+    for i in 1:length(Tar.prob.u0)
+        # can add l2std[i] for u[i]
+        # for u[i] ith vector must be added to dataset,nn[1,:] is the dx in lotka_volterra
+        L2logprob += logpdf(MvNormal(nn[i, :], Tar.l2std[i]), Tar.dataset[i])
+    end
+    return L2logprob
+end
+
+# priors for NN parameters + ODE constants
+function priorweights(Tar::LogTargetDensity, θ)
+    allparams = Tar.priors
+    # ode parameters
+    invparams = allparams[2:length(allparams)]
+    stdinv = [invparam[2] for invparam in invparams]
+    # nn weights
+    nnwparams = allparams[1]
+    stdw = nnwparams[2]
+    prisw = nnwparams[1] .* ones(length(θ) - Tar.extraparams)
+
+    if Tar.extraparams > 0
+        return (logpdf(MvNormal(zeros(Tar.extraparams), stdinv),
```

Comment:
`MvNormal` expects variance not stddev

Reply:
the Distributions.MvNormal() ends up creating the covariance by squaring the stddev.
```
julia> MvNormal([1,2],[1,3])
MvNormal{Int64, PDMats.PDiagMat{Int64, Vector{Int64}}, Vector{Int64}}(
dim: 2
μ: [1, 2]
Σ: [1 0; 0 9]
)
```